### ISSUE TITLE
Upgrade of Insights 2022.3 does not work

### ISSUE THREAD

**Issue by alexanderwalkowski** [Esri/arcgis-powershell-dsc#445]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request

### Module Version

* 4.0.2

### Affected Resource(s)

* Insights for ArcGIS

### Configuration Files

```
{
    "AllNodes": [
        {
            "NodeName": "ServerHost",
            "Role": [
                "Server"
            ]
        },
        {
            "NodeName": "PortalHost",
            "Role": [
                "Portal"
            ]
        }
    ],
    "ConfigData": {
        "Version": "10.9.1",
        "OldVersion": "10.9.1",
        "ServerContext": "arcgis",
        "PortalContext": "portal4arcgis",
        "ServerRole": "GeneralPurposeServer",
        "InsightsVersion": "2022.3",
        "OldInsightsVersion": "2022.1.1",
        "DownloadSetups": false,
        "DownloadPatches":false,
        "Credentials": {
            "ServiceAccount": {
                "UserName": "user",
                "Password": "password",
                "IsDomainAccount": true,
                "IsMSAAccount": false
            }
        },        
        "Server": {
            "LicenseFilePath": "C:\\data\\installation\\ArcGISGISServerStandard_ArcGISServer_1144804.prvc",
            "Installer": {
                "Path": "C:\\data\\installation\\ArcGIS_Server_Windows_1091_180041.exe",
                "InstallDir": "C:\\bin\\ArcGIS\\Server",
                "InstallDirPython": "C:\\bin\\Python27",
                "EnableArcMapRuntime": true,
                "EnableDotnetSupport": true,
                "PatchesDir": "C:\\data\\installation\\arcgisPatches",
                "PatchInstallOrder": [
                    "ArcGIS-1091-S-SP-Patch.msp",
                    "ArcGIS-1091-S-MR-PatchB.msp",
                    "ArcGIS-1091-S-Log4j-PatchB.msp",
                    "ArcGIS-1091-S-SQ-Patch.msp",
                    "ArcGIS-1091-S-UNDM1-Patch.msp",
                    "ArcGIS-1091-S-VRPG-Patch.msp",
                    "MappingandChartingServer1091Update1.msp",
                    "ArcGIS-1091-S-MPS-Patch.msp",
                    "ArcGIS-1091-S-SEC2022U1-PatchB.msp",
                    "MappingandChartingServer1091Update2.msp",
                    "ArcGIS-1091-S-UNDM2-Patch.msp",
                    "ArcGIS-1091-S-SEC2022U2-Patch.msp",
                    "MappingandChartingServer1091Update3.msp",
                    "ArcGIS-1091-S-UNDM3-Patch.msp",
                    "ArcGIS-1091-S-GSQ-Patch.msp"
                ]
            },
            "ServerDirectoriesRootLocation": "C:\\arcgisserver\\directories",
            "ConfigStoreLocation": "C:\\arcgisserver\\config-store",
            "LocalRepositoryPath": "C:\\arcgisserver\\local",
            "ServerLogsLocation": "C:\\arcgisserver\\logs",
            "ServerDirectories": [
                {
                    "name": "arcgiscache",
                    "physicalPath": "C:\\arcgisserver\\directories\\arcgiscache",
                    "directoryType": "CACHE",
                    "cleanupMode": "NONE",
                    "maxFileAge": 0
                },
                {
                    "name": "arcgisjobs",
                    "physicalPath": "C:\\arcgisserver\\directories\\arcgisjobs",
                    "directoryType": "JOBS",
                    "cleanupMode": "TIME_ELAPSED_SINCE_LAST_MODIFIED",
                    "maxFileAge": 360
                },
                {
                    "name": "arcgisoutput",
                    "physicalPath": "C:\\arcgisserver\\directories\\arcgisoutput",
                    "directoryType": "OUTPUT",
                    "cleanupMode": "TIME_ELAPSED_SINCE_LAST_MODIFIED",
                    "maxFileAge": 10
                },
                {
                    "name": "arcgissystem",
                    "physicalPath": "C:\\arcgisserver\\directories\\arcgissystem",
                    "directoryType": "SYSTEM",
                    "cleanupMode": "NONE",
                    "maxFileAge": 0
                }
            ],
            "PrimarySiteAdmin": {
                "UserName": "siteadmin",
                "Password": "siteadmin"
            },
            "Extensions": {
                "MappingChartingSolution": {
                    "Installer": {
                        "Path": "C:\\data\\installation\\ArcGIS_Mapping_and_Charting_Solutions_Server_1091_180544.exe",
                        "InstallDir": "C:\\bin\\ArcGIS\\MappingChartingSolution"
                    },
                    "Features": [
                        "Maritime"
                    ]
                }
            }
        },
        "Portal": {
            "LicenseFilePath": "C:\\data\\installation\\ArcGIS_Enterprise_Portal_1091_367221_20211130.json",
            "PortalLicenseUserTypeId": "creatorUT",
            "Installer": {
                "Path": "C:\\data\\installation\\Portal_for_ArcGIS_Windows_1091_180052.exe",
                "WebStylesPath": "C:\\data\\installation\\Portal_for_ArcGIS_Web_Styles_Windows_1091_180053.exe",
                "InstallDir": "C:\\bin\\ArcGIS\\Portal",
                "ContentDir": "C:\\arcgisportal",
                "PatchesDir": "C:\\data\\installation\\arcgisPatches",
                "PatchInstallOrder": [
                    "ArcGIS-1091-PFA-Log4j-PatchB.msp",
                    "ArcGIS-1091-PFA-QCS-Patch.msp",
                    "ArcGIS-1091-PFA-SEC2022U1-Patch.msp",
                    "ArcGIS-1091-PFA-SEC2022U2-PatchB.msp",
                    "ArcGIS-1091-PFA-ESFD-Patch.msp"
                ]
            },
            "ContentDirectoryLocation": "C:\\arcgisportal\\content",
            "EnableAutomaticAccountCreation": false,
            "DefaultRoleForUser": "viewer",
            "DefaultUserLicenseTypeIdForUser": "viewerUT",
            "PortalAdministrator": {
                "UserName": "user",
                "Email": "mail",
                "Password": "password",
                "SecurityQuestionIndex": 1,
                "SecurityAnswer": "vanilla"
            },
            "EmailSettings": {
                "SMTPServerAddress": "host",
                "From": "mail",
                "Label": "Portal Admin",
                "AuthenticationRequired": true,
                "SMTPPort": "port",
                "UserName": "user",
                "Password": "password",
                "EncryptionMethod": "TLS"
            }
        },
        "Insights": {
            "Installer": {
                "Path": "C:\\data\\installation\\ArcGIS_Insights_Windows_2022_3_184093.exe"
            }
        }
    }
}
```

### Expected Behavior

* Insights for ArcGIS should be updated
* my expectation would be, that I just need to specify the oldInsightsVersion, InsightsVersion and the Insights Block in the ConfigData section, since I just want to update Insights and do not need any changes on server and portal itself.

### Actual Behavior

* no installation task of insights is logged
* version.txt in the installation folder of insights on the portal and server hosts, show the old version

### Steps to Reproduce

* execute the following command referencing the above mentioned configuration file.
`Invoke-ArcGISConfiguration -ConfigurationParametersFile ./Deployment-upgrageInsights.json -Mode Upgrade -EnableMSILogging -DebugSwitch`

**Comment by alexanderwalkowski** [comment]
I tried the installation (Install) mode instead of upgrade, with the following configuration file:
```
{
    "AllNodes": [
        {
            "NodeName": "ServerHost",
            "Role": [
                "Server"
            ]
        },
        {
            "NodeName": "PortalHost",
            "Role": [
                "Portal"
            ]
        }
    ],
    "ConfigData": {
        "Version": "10.9.1",       
        "ServerContext": "arcgis",
        "PortalContext": "portal4arcgis",
        "ServerRole": "GeneralPurposeServer",
        "InsightsVersion": "2022.3",        
        "DownloadSetups": false,
        "DownloadPatches":false,
        "Credentials": {
            "ServiceAccount": {
                "UserName": "user",
                "Password": "password",
                "IsDomainAccount": true,
                "IsMSAAccount": false
            }
        },        
        "Insights": {
            "Installer": {
                "Path": "C:\\data\\installation\\ArcGIS_Insights_Windows_2022_3_184093.exe"
            }
        }
    }
}
```

But I get the following errors:


```
Dot Sourcing the Configuration:- ArcGISInstall
PSDesiredStateConfiguration\Node : Es ist nicht möglich, eine Methode für einen Ausdruck aufzurufen, der den NULL hat.
--> TRANSLATION: it is not possible, to invoke a method on a null value expression.
In C:\Program Files\WindowsPowerShell\Modules\ArcGIS\4.0.2\Configurations-OnPrem\ArcGISInstall.ps1:27 Zeichen:5
+     Node $AllNodes.NodeName {
+     ~~~~
    + CategoryInfo          : InvalidOperation: (:) [PSDesiredStateConfiguration\node], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : InvokeMethodOnNull,PSDesiredStateConfiguration\node
WARNUNG: Es wird davon abgeraten, Domänenanmeldeinformationen für den Knoten 'portalHost' zu verwenden. Um die Warnung zu unterdrücken, können Sie Ihren DSC-Konfigurationsdaten für Knoten 'DEV0308W' eine Eigenschaft namens 
'PSDscAllowDomainUser' mit dem Wert '$true' hinzufügen.
Kompilierungsfehler bei der Verarbeitung der Konfiguration "ArcGISInstall". Überprüfen Sie die im Fehlerdatenstrom gemeldeten Fehler, und ändern Sie den Konfigurationscode entsprechend.
In C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules\PSDesiredStateConfiguration\PSDesiredStateConfiguration.psm1:3917 Zeichen:5
+     throw $ErrorRecord
+     ~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (ArcGISInstall:String) [], InvalidOperationException
    + FullyQualifiedErrorId : FailToProcessConfiguration
```

**Comment by cameronkroeker** [comment]
Hi @alexanderwalkowski,

When using `-Mode Upgrade` the json configuration file will need to include both `ConfigData.OldVersion`, and `ConfigData.OldInsightsVersion` attributes.

And when using `-Mode Install` the json configuration file will need to include `ConfigData.Portal` block and `ConfigData.Server` block.

For examle:

```
{
    "AllNodes": [
        {
            "NodeName": "host1",
            "Role": [
                "Server"
            ]
        },
        {
            "NodeName": "host2",
            "Role": [
                "Portal"
            ]
        }
    ],
    "ConfigData": {
        "Version": "10.9.1",
        "OldVersion": "10.9.1",
        "ServerContext": "arcgis",
        "PortalContext": "portal4arcgis",
        "ServerRole": "GeneralPurposeServer",
        "InsightsVersion": "2022.3",
        "OldInsightsVersion": "2022.1.1",
        "DownloadSetups": false,
        "DownloadPatches":false,
        "Credentials": {
            "ServiceAccount": {
                "UserName": "username",
                "Password": "password",
                "IsDomainAccount": true,
                "IsMSAAccount": false
            }
        },
       
        "Server": {
            "LicenseFilePath": "C:\\data\\installation\\ArcGISGISServerStandard_ArcGISServer_1111807.prvc",
            "Installer": {
                "Path": "C:\\data\\installation\\ArcGIS_Server_Windows_1091_180041.exe",
                "InstallDir": "C:\\bin\\ArcGIS\\Server",
                "InstallDirPython": "C:\\bin\\Python27",
                "EnableArcMapRuntime": true,
                "EnableDotnetSupport": true,
                "PatchesDir": "C:\\data\\installation\\arcgisPatches",
                "PatchInstallOrder": [
                    "ArcGIS-1091-S-SP-Patch.msp",
                    "ArcGIS-1091-S-MR-PatchB.msp",
                    "ArcGIS-1091-S-Log4j-PatchB.msp",
                    "ArcGIS-1091-S-SQ-Patch.msp",
                    "ArcGIS-1091-S-UNDM1-Patch.msp",
                    "ArcGIS-1091-S-VRPG-Patch.msp",
                    "MappingandChartingServer1091Update1.msp",
                    "ArcGIS-1091-S-MPS-Patch.msp",
                    "ArcGIS-1091-S-SEC2022U1-PatchB.msp",
                    "MappingandChartingServer1091Update2.msp",
                    "ArcGIS-1091-S-UNDM2-Patch.msp",
                    "ArcGIS-1091-S-SEC2022U2-Patch.msp",
                    "MappingandChartingServer1091Update3.msp",
                    "ArcGIS-1091-S-UNDM3-Patch.msp",
                    "ArcGIS-1091-S-GSQ-Patch.msp"
                ]
            },
            "ServerDirectoriesRootLocation": "C:\\arcgisserver\\directories",
            "ConfigStoreLocation": "C:\\arcgisserver\\config-store",
            "LocalRepositoryPath": "C:\\arcgisserver\\local",
            "ServerLogsLocation": "C:\\arcgisserver\\logs",
            "ServerDirectories": [
                {
                    "name": "arcgiscache",
                    "physicalPath": "C:\\arcgisserver\\directories\\arcgiscache",
                    "directoryType": "CACHE",
                    "cleanupMode": "NONE",
                    "maxFileAge": 0
                },
                {
                    "name": "arcgisjobs",
                    "physicalPath": "C:\\arcgisserver\\directories\\arcgisjobs",
                    "directoryType": "JOBS",
                    "cleanupMode": "TIME_ELAPSED_SINCE_LAST_MODIFIED",
                    "maxFileAge": 360
                },
                {
                    "name": "arcgisoutput",
                    "physicalPath": "C:\\arcgisserver\\directories\\arcgisoutput",
                    "directoryType": "OUTPUT",
                    "cleanupMode": "TIME_ELAPSED_SINCE_LAST_MODIFIED",
                    "maxFileAge": 10
                },
                {
                    "name": "arcgissystem",
                    "physicalPath": "C:\\arcgisserver\\directories\\arcgissystem",
                    "directoryType": "SYSTEM",
                    "cleanupMode": "NONE",
                    "maxFileAge": 0
                }
            ],
            "PrimarySiteAdmin": {
                "UserName": "siteadmin",
                "Password": "siteadmin"
            },
            "Extensions": {
                "MappingChartingSolution": {
                    "Installer": {
                        "Path": "C:\\data\\installation\\ArcGIS_Mapping_and_Charting_Solutions_Server_1091_180544.exe",
                        "InstallDir": "C:\\bin\\ArcGIS\\MappingChartingSolution"
                    },
                    "Features": [
                        "Maritime"
                    ]
                }
            }
        },
        "Portal": {
            "LicenseFilePath": "C:\\data\\installation\\ArcGIS_Enterprise_Portal_1091_367221_20211130.json",
            "PortalLicenseUserTypeId": "creatorUT",
            "Installer": {
                "Path": "C:\\data\\installation\\Portal_for_ArcGIS_Windows_1091_180052.exe",
                "WebStylesPath": "C:\\data\\installation\\Portal_for_ArcGIS_Web_Styles_Windows_1091_180053.exe",
                "InstallDir": "C:\\bin\\ArcGIS\\Portal",
                "ContentDir": "C:\\arcgisportal",
                "PatchesDir": "C:\\data\\nstallation\\arcgisPatches",
                "PatchInstallOrder": [
                    "ArcGIS-1091-PFA-Log4j-PatchB.msp",
                    "ArcGIS-1091-PFA-QCS-Patch.msp",
                    "ArcGIS-1091-PFA-SEC2022U1-Patch.msp",
                    "ArcGIS-1091-PFA-SEC2022U2-PatchB.msp",
                    "ArcGIS-1091-PFA-ESFD-Patch.msp"
                ]
            },
            "ContentDirectoryLocation": "C:\\arcgisportal\\content",
            "EnableAutomaticAccountCreation": false,
            "DefaultRoleForUser": "viewer",
            "DefaultUserLicenseTypeIdForUser": "viewerUT",
            "PortalAdministrator": {
                "UserName": "portaladmin",
                "Email": "mail.de",
                "Password": "password",
                "SecurityQuestionIndex": 1,
                "SecurityAnswer": "answer"
            },
            "EmailSettings": {
                "SMTPServerAddress": "mailserver",
                "From": "noreply@mail.de",
                "Label": "Portal Admin",
                "AuthenticationRequired": true,
                "SMTPPort": "587",
                "UserName": "username",
                "Password": "password",
                "EncryptionMethod": "TLS"
            }
        },
        "DataStore": {
            "ContentDirectoryLocation": "C:\\arcgisdatastore",
            "EnableFailoverOnPrimaryStop": false,
            "EnablePointInTimeRecovery": false,
            "Installer": {
                "Path": "C:\\data\\installation\\ArcGIS_DataStore_Windows_1091_180054.exe",
                "InstallDir": "C:\\bin\\ArcGIS\\DataStore",
                "PatchesDir": "C:\\data\\installation\\arcgisPatches",
                "PatchInstallOrder": [
                    "ArcGIS-1091-DS-Log4j-Patch.msp",
                    "ArcGIS-1091-DS-DE-Patch.msp"
                ]
            }
        },
        "Insights": {
            "Installer": {
                "Path": "C:\\data\\installation\\ArcGIS_Insights_Windows_2022_3_184093.exe"
            }
        }
    }
}
```

Thanks,
Cameron K.

**Comment by alexanderwalkowski** [comment]
Hi @cameronkroeker,
thanks for clarifying the dependencies between the `-Mode` parameter and the required blocks in the configuration date. It might be helpfull to include this in the wiki.

With your clarification in mind, I double checked my configuration file, and the initial issue still appears. Just to make sure I understood you right, let me sumarize my setup:
* Test 1
** mode `upgrade`
** configuration includes oldVersion=10.9.1 version=10.9.1, insightsVersion and oldInsightsVersion
** configuration includes server and portal block
** result
*** dsc script exits without failure, but no installation steps concerning insights are logged and the version.txt in the installation folder is not updated
* Test 2
** mode `upgrade`
** configuration includes oldVersion=10.9.1 version=10.9.1, insightsVersion and oldInsightsVersion
** configuration does not include server and portal block
** result
*** tbd
* Test 3
** mode `install`
** configuration includes only insightsVersion=2022.3
** configuration includes configData.server and configData.portal block
** result
*** tbd

**Comment by cameronkroeker** [comment]
Hi @alexanderwalkowski,

The `-Mode Upgrade` will first uninstall the previous version of Insights (`ConfigData.OldInsightsVersion`) and install the new version (`ConfigData.InsightsVersion`). This ideally should be used when upgrading both ArcGIS Enterprise and ArcGIS Insights. Though it should technically work when only upgrading ArcGIS Insights because the module is idempotent and will skip upgrading ArcGIS Enterprise.

If only upgrading ArcGIS Insights, then perhaps it makes more sense to use `-Mode Install` (as you did in Test 3). This will do an "in-place" upgrade of Insights. This is fine for newer versions, however I believe 10.7 and prior an "in-place" upgrade was not always supported.

https://doc.arcgis.com/en/insights/2022.1/get-started/install-insights.htm#ESRI_SECTION1_E87D9F78C1BD469DA9F75AF040F44B50

I suspect that Test 1 did not fully work because `ConfigData.Version` and `ConfigData.OldVersion` matched. I would be interested in seeing this test ran again but with `ConfigData.OldVersion`  set to something prior to 10.9.1, for example 10.8.1.

Your assessment of Test 2 is accurate. My apologies, I meant the `ConfigData.Portal` and `ConfigData.Server` block should contain the original attributes and values (usernames/passwords/install dir etc).

Thanks,
Cameron K.

**Comment by alexanderwalkowski** [comment]
Hi @cameronkroeker,
Sorry, but I am not able to rerun test 1, since the installation was successful in test 3.

In my opinion it would be helpful, to enhance the documentation
* concerning the dependencies between the configData blocks and the mode
* by an example how to update Insights without updating the enterprise version

kind regards,
Alexander

**Comment by cameronkroeker** [comment]
Hi @alexanderwalkowski,

No worries, I was able to verify Test 1. Test 1 didn't work because of 2 reasons, first the `ConfigData.OldVersion` attribute value needs to be lower than `ConfigData.Version`, and second the full Enterprise stack (Server\Portal\DataStore\WebAdaptor) needs to be specified in the Role sections for it to invoke the Insights upgrade.

So what we have learned from this is, when upgrading both ArcGIS Enterprise and Insights then it's best to use `-Mode Upgrade`, and when only upgrading ArcGIS Insights then its best to use `-Mode Install`. I also do agree with you that there should be documentation that provides these details.

Thanks!
